### PR TITLE
add an temporal graph example

### PR DESCRIPTION
**Pull request by CongWeilin**
*Issue #, if available:*

*Description of changes:*

Add a temporal graph learning example: train a custimized TGAT on snapshot-based temporal graph build from the original MAG dataset.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by CongWeilin** [comment]
Please let me know how we would like to organize the `README.md` file, and whether we want the following information added to `README.md` file:

- A high-level description on our customized TGAT model implementation.
- Explaination on how we do temporal sampling, why current framework cannot support too many timestamps.
- Details on dataset MAG, and high-level description on how the graph is constructed.

File: examples/temporal_graph_learning/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Customized TGAT model for snapshot-based temporal graph learning
+
+This folder contains python codes that demonstrate how to leverage the GraphStorm Framework (GSF) for production level snapshot-based temporal graph data.
+This example uses the TGAT as GNN model and the MAG graph data.
```
Comment (zheng-da) [inline examples/temporal_graph_learning/README.md:4]: i thought you wanted to use GDELT.

File: examples/temporal_graph_learning/README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+The following commands assume user in directory `graphstorm/examples/temporal_graph_learning/`.
+
+**Step 1: Prepare the MAG dataset for using the GraphStorm.**
+
+Download dataset
+```
+mkdir ./DATA
+wget -P ./DATA/GDELT https://s3.us-west-2.amazonaws.com/dgl-data/dataset/tgl/GDELT/node_features.pt
+wget -P ./DATA/GDELT https://s3.us-west-2.amazonaws.com/dgl-data/dataset/tgl/GDELT/labels.csv
+wget -P ./DATA/GDELT https://s3.us-west-2.amazonaws.com/dgl-data/dataset/tgl/GDELT/edges.csv
```
Comment (zheng-da) [inline examples/temporal_graph_learning/README.md:17]: is it GDELT or MAG?

File: examples/temporal_graph_learning/gen_graph.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+def create_dgl_graph():
+    # load dataset
+    data = 'MAG'
+    df = pd.read_csv('./DATA/{}/edges.csv'.format(data))
+    num_nodes = max(int(df['src'].max()), int(df['dst'].max())) + 1
+    print('num_nodes: ', num_nodes)
+    
+    # Extract graph structure 
+    # We only consider edge types from 100 to 109
+    timestamp = df['time'].values
```
Comment (zheng-da) [inline examples/temporal_graph_learning/gen_graph.py:19]: the timestamp is on the edges. is this MAG or GDELT?

File: examples/temporal_graph_learning/gen_graph.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    mask = np.logical_and(timestamp >= 100, timestamp < 110)
+    
+    timestamp = timestamp[mask]
+    src_nodes = df['src'].values[mask]
+    dst_nodes = df['dst'].values[mask]
```
Comment (zheng-da) [inline examples/temporal_graph_learning/gen_graph.py:24]: you only extracted a small number of edges?

File: examples/temporal_graph_learning/gen_graph.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    g.nodes['paper'].data['test_mask'] = torch.from_numpy(test_mask).to(torch.uint8)
+    g.nodes['paper'].data['feat'] = node_feats[unique_nodes].float()
+    
+    return g
+
+
+g = create_dgl_graph()
+print(g)
+print(g.is_homogeneous)
+dgl.distributed.partition_graph(g, graph_name='MAG_Temporal', num_parts=1, out_path='./DATA/')
```
Comment (zheng-da) [inline examples/temporal_graph_learning/gen_graph.py:123]: instead of creating the graph yourself, i think it's better to use GraphStorm's graph construction pipeline to create this graph. In this way, users can know better how to construct a temporal graph with GraphStorm.

File: examples/temporal_graph_learning/gnn_encoder.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        )
+
+    def forward(self, blocks, inputs):
+        h = to_per_field_nfeats(inputs, self.node_fields)
+        for layer, block in zip(self.layers, blocks):
+            h = layer(block, h)
+        h = average_over_fields(h)
+        return h
+
+class RelGraphConvLayer(nn.Module):
```
Comment (zheng-da) [inline examples/temporal_graph_learning/gnn_encoder.py:52]: is this RGAT?

Comment (zheng-da) [reply examples/temporal_graph_learning/gnn_encoder.py:52]: please provide docstrings for every class.

File: examples/temporal_graph_learning/gen_graph.py (hunk 3 above)
Comment (zheng-da) [reply examples/temporal_graph_learning/gen_graph.py:19]: instead of filtering out edges, you can potentially discretize the timestamps. For example, putting 1-10 as timestamp 1, 11-20 as timestamp 2, etc.

File: examples/temporal_graph_learning/gnn_encoder.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+from model_utils import merge_multi_blocks, get_merge_canonical_etype_mapping
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from dgl.nn import GATConv
+
+class RelationalGraphEncoder(gsmodel.gnn_encoder_base.GraphConvEncoder):
```
Comment (zheng-da) [inline examples/temporal_graph_learning/gnn_encoder.py:19]: where do you put temporal encoding?

Comment (zheng-da) [reply examples/temporal_graph_learning/gnn_encoder.py:19]: we need to have one place to explain TGAT and how TGAT is implemented in this code base.

File: examples/temporal_graph_learning/README.md (hunk 1 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/README.md:4]: For this example, I am using a modified version of MAG from https://github.com/amazon-science/tgl.

File: examples/temporal_graph_learning/gen_graph.py (hunk 3 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/gen_graph.py]: Yes, timestamp is on edges.

For this example, since the original MAG dataset is very large, I decided to only consider 10 timestamps instead of discretizing timestamp and using the full graph. 

When using only 10 timestamps, the DGL graph is already 100GB. I tried discretizing the timestamps, it takes too much time and the graph is too big. I think it is better to keep example data small so that user can easily test them.

File: examples/temporal_graph_learning/gen_graph.py (hunk 4 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/gen_graph.py]: Yes, I first select all edges with timestamp from 100 - 110 (those are most dense graph with many labeled nodes).

Then, I only keep nodes that appears during 100 - 110, and constructed a subsampled version of MAG dataset. By doing so, we won't have nodes that doesn't have any neighbors.

File: examples/temporal_graph_learning/gen_graph.py (hunk 5 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/gen_graph.py]: I updated the `gen_graph.py` and `README.md` to use GraphStorm's graph construction pipeline. Please check.

File: examples/temporal_graph_learning/gnn_encoder.py (hunk 7 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/gnn_encoder.py]: I added a paragraph in `README.md` to explain how we utilize temporal information based on the GraphStorm's  RGAT.

File: examples/temporal_graph_learning/gnn_encoder.py (hunk 6 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/gnn_encoder.py]: I added docstring to each function

I also update function/class name accordingly.

File: examples/temporal_graph_learning/input_encoder.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        feat_size,
+        dropout=0.0,
+    ):
+        super(NodeEncoderInputLayer, self).__init__(g)
+        self.activation = nn.LeakyReLU()
+        self.dropout = nn.Dropout(dropout)
+        self.out_size = out_size
+        self.feat_size = feat_size
+
+        self.fields = get_unique_nfields(g.etypes)
```
Comment (zheng-da) [inline examples/temporal_graph_learning/input_encoder.py:48]: in this example, do you assume that there is only one relation type originally? the code may only work on a temporal homogeneous graph. we need to make it work for temporal heterogeneous graphs as well.

Comment (CongWeilin) [reply examples/temporal_graph_learning/input_encoder.py:48]: This code works for hetergenious graph with multiple relation types.

Comment (CongWeilin) [reply examples/temporal_graph_learning/input_encoder.py:48]: This code was based on Cramer's code, I think I didn't make much changes on the number of relation types.

File: examples/temporal_graph_learning/generate_launch_script_nc.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
+import yaml
```
Comment (zheng-da) [inline examples/temporal_graph_learning/generate_launch_script_nc.py:1]: i think you can just provide a yaml file directly. please see other examples.

File: examples/temporal_graph_learning/gnn_encoder.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        """Forward computation.
+
+        Parameters
+        ----------
+        blocks: DGL MFGs
+            Sampled subgraph in DGL MFG
+        inputs: dict[str, torch.Tensor]
+            Input node feature for each node type.
+        """
+        h = to_per_field_nfeats(inputs, self.node_fields)
```
Comment (zheng-da) [inline examples/temporal_graph_learning/gnn_encoder.py:71]: this assumes a format that stores the node features at different timestamps. the format needs to be explicitly explained in the readme.

File: examples/temporal_graph_learning/gnn_encoder.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        self.loop_weights = nn.Linear(in_feat, out_feat, bias=True)
+
+    def forward(self, g, inputs):
+        """Forward computation. Residual connection is used.
+
+        Parameters
+        ----------
+        g: DGL MFGs
+            Sampled subgraph in DGL MFG
+        inputs: dict[str, torch.Tensor]
```
Comment (zheng-da) [inline examples/temporal_graph_learning/gnn_encoder.py:112]: is this true? i thought `inputs` is a dict of dicts of tensors.

File: examples/temporal_graph_learning/model_utils.py [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+    Returns
+    -------
+        list of int: list of sorted number suffix of each relation name
+    """
+    field_ids = [rel_field_map(rel_name) for rel_name in rel_names]
+    field_ids = list(set(field_ids))
+    field_ids.sort()
+    return field_ids
+
+def get_merge_canonical_etype_mapping(etypes):
```
Comment (zheng-da) [inline examples/temporal_graph_learning/model_utils.py:162]: i think you should rename it. it doesn't suggest this is to get the mapping for temporal aggregation. to make the name shorter, you can remove canonical.

File: examples/temporal_graph_learning/input_encoder.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        # field_embeds act as temporal encoding
+        self.field_embeds = nn.ParameterDict()
+        for t in self.fields:
+            self.field_embeds[f"{t}_feat"] = get_trainable_params(1, out_size)
```
Comment (zheng-da) [inline examples/temporal_graph_learning/input_encoder.py:52]: this temporal encoding is basically assign different learnable embeddings for different timestamps? if so, how can the model understand the order in the sequence?

File: examples/temporal_graph_learning/gnn_encoder.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+                canonical_etypes[1],
+                canonical_etypes[2],
+            )
+            gc_layer = GATConv(
+                in_feat,
+                out_feat // num_heads,
+                num_heads,
+                bias=False,
+                allow_zero_in_degree=True,
+            )
```
Comment (zheng-da) [inline examples/temporal_graph_learning/gnn_encoder.py:168]: this implementation also uses the model parameters to distinguish different timestamps.

File: python/graphstorm/config/argument.py [hunk 15]
```diff
@@ -1236,6 +1237,16 @@ def target_ntype(self):
             "Must provide the target ntype through target_ntype"
         return self._target_ntype
 
+    @property
+    def eval_target_ntype(self):
```
Comment (zheng-da) [inline python/graphstorm/config/argument.py:1241]: is this a new parameter?

Comment (zheng-da) [reply python/graphstorm/config/argument.py:1241]: this modification is better to be in another PR. the PR is to add an example. it should not modify the core code.

File: examples/temporal_graph_learning/generate_launch_script_nc.py (hunk 9 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/generate_launch_script_nc.py]: Done.

File: examples/temporal_graph_learning/gnn_encoder.py (hunk 10 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/gnn_encoder.py:71]: I short paragraph is added to README's **Implementation** section.

File: examples/temporal_graph_learning/gnn_encoder.py (hunk 11 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/gnn_encoder.py]: Fixed. Change to dict[str, dict[str, torch.Tensor]].

File: examples/temporal_graph_learning/model_utils.py (hunk 12 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/model_utils.py:162]: I renamed it to `get_temporal_ordered_etypes`

File: examples/temporal_graph_learning/input_encoder.py (hunk 13 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/input_encoder.py:52]: I was expecting neural networks can learn this sequence from training data, because such sequential information could be learned from our temporal aggregation strategy:

hidden embeddings at day 1 is only computed on day1's input features, 

hidden embeddings at day 2 is only computed on day1 & day2's input features, ...

File: examples/temporal_graph_learning/gnn_encoder.py (hunk 14 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/gnn_encoder.py:168]: No. `canonical_etypes` is returned by `get_unique_etype_triplet`, which removed the timestamp information on the edge type.

In this example, we have
```
self.conv = {
       'paper-cite-paper': dgl.nn.GATConv
}
```

File: python/graphstorm/config/argument.py (hunk 15 above)
Comment (CongWeilin) [reply python/graphstorm/config/argument.py]: I remove those changes from this PR. I will create a new PR later.

File: examples/temporal_graph_learning/README.md [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+```
+
+**Step 2: Pre-processing the raw data into GraphStorm/DGL's data structure**
+
+Run the command to convert the data into the required raw format:
+```
+python3 gen_graph.py
+```
+
+Once successful, the command will generate a set of folders and files under the `./DATA` folder, which contains pre-processed graph data recognizable by the GraphStorm's graph construction tool.
```
Comment (zhjwy9343) [inline examples/temporal_graph_learning/README.md:26]: Can we explain what the data is split by time and what the outcomes of the `gen_graph.py` script? 

Because temporal information is the key concept in this example, it is better to have some explanation about how we construct data to fit the TGAT. thanks

File: examples/temporal_graph_learning/model_nc.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        The graph used in training and testing
+    config: GSConfig
+        Configurations
+
+    Returns
+    -------
+    GSgnnNodeModel : The GNN model.
+    """
+
+    model = gs.model.GSgnnNodeModel(config.alpha_l2norm)
```
[the remark this answers is not in the document]
Comment (CongWeilin) [reply examples/temporal_graph_learning/model_nc.py:27]: In this implementation, we are actually using the exact `GSgnnNodeModel` but setting a customized gnn_encoder. We are still using the forward and predict function in `GSgnnNodeModel`. 

In this case, we might prefer `GSgnnNodeModel` than `GSgnnNodeModelBase` , as we are not creating an entirely new model from scratch.

File: examples/temporal_graph_learning/README.md (hunk 16 above)
Comment (CongWeilin) [reply examples/temporal_graph_learning/README.md]: I made some updates to README.md please check. Please let me know if need to include more information.

File: examples/temporal_graph_learning/gnn_encoder.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+        Number of attention heads
+    """
+    def __init__(self, g, in_feat, out_feat, num_heads):
+        super(TemporalRelationalGraphConv, self).__init__()
+        self.in_feat = in_feat
+        self.out_feat = out_feat
+        self.node_fields = get_unique_nfields(g.etypes)
+
+        self.conv = nn.ModuleDict()
+        for canonical_etypes in get_unique_etype_triplet(g.canonical_etypes):
```
Comment (zheng-da) [inline examples/temporal_graph_learning/gnn_encoder.py:156]: can you write comment here to explain that the etype doesn't contain timestamp? it means that all timestamps share the same message passing model.

File: examples/temporal_graph_learning/input_encoder.py (hunk 13 above)
Comment (zheng-da) [reply examples/temporal_graph_learning/input_encoder.py:52]: can you write comments to explain how temporal encoding works in this example code.

**Review by zheng-da — APPROVED** [review]
